src/base-subscription.ts

```diff
@@ -0,0 +1,281 @@
+import { Logger } from './logger';
+import { TokenProvider } from './token-provider';
+import { XhrReadyState, NetworkError, ErrorResponse, Headers } from "./base-client";
+
+export enum SubscriptionState {
+    UNOPENED = 0, // haven't called xhr.send()
+    OPENING,      // called xhr.send(); not yet received response headers
+    OPEN,         // received response headers; called onOpen(); expecting message
+    ENDING,       // received EOS message; response not yet finished
+    ENDED         // called onEnd() or onError(err)
+}
+
+export interface SubscriptionEvent {
+    eventId: string;
+    headers: Headers;
+    body: any;
+}
+
+export interface SubscribeOptions {    
+    path: string;
+    headers: Headers;
+    
+    logger: Logger;
+    
+    onOpen?: () => void;
+    onEvent?: (event: SubscriptionEvent) => void;
+    onEnd?: () => void;
+    onError?: (error: Error) => void;
+    onRetry?: () => void;
+}
+
+export class BaseSubscription {
+    
+    private state: SubscriptionState = SubscriptionState.UNOPENED;
+    
+    constructor(
+        private xhr: XMLHttpRequest,
+        private options: SubscribeOptions
+    ){  
+        //Apply headers    
+        for (let key in options.headers) {
+            xhr.setRequestHeader(key, options.headers[key]);
+        }
+                
+        xhr.onreadystatechange = () => {
+            switch(this.xhr.readyState) {
+                case XhrReadyState.UNSENT:
+                case XhrReadyState.OPENED:
+                case XhrReadyState.HEADERS_RECEIVED:
+                this.assertStateIsIn(SubscriptionState.OPENING);
+                break;
+                
+                case XhrReadyState.LOADING:
+                this.onLoading();
+                break;
+                
+                case XhrReadyState.DONE:
+                this.onDone();
+                break; 
+            }
+        } 
+    }
+
+    open(jwt?: string) {
+        if (this.state !== SubscriptionState.UNOPENED) {
+            throw new Error("Called .open() on Subscription object in unexpected state: " + this.state);
+        }
+
+        this.state = SubscriptionState.OPENING;
+
+        if (jwt) {
+            this.xhr.setRequestHeader("authorization", `Bearer ${jwt}`);
+        }
+
+        this.xhr.send();
+    }
+
+    unsubscribe(err: any): void {
```

Comment:
Ok, I'm not sure about handling failure cases in unsubscribe. It's an explicit call from the consumer to terminate the subscription - it sounds like a scenario that shouldn't result with an error.